src/Storages/MergeTree/MutateTask.cpp

```diff
@@ -1895,13 +2070,82 @@ class MutateSomePartColumnsTask : public IExecutableTask
 
             ctx->projections_to_build = std::vector<ProjectionDescriptionRawPtr>{ctx->projections_to_recalc.begin(), ctx->projections_to_recalc.end()};
 
+            for (const auto & mutation_context : ctx->projection_mutation_contexts)
+            {
+                chassert (mutation_context->mutating_pipeline_builder.initialized());
+
+                mutation_context->new_part->getDataPartStorage().createDirectories();
+
+                /// Create hardlinks for unchanged files
+                for (auto it = mutation_context->source_part->getDataPartStorage().iterate(); it->isValid(); it->next())
+                {
+                    if (mutation_context->files_to_skip.contains(it->name()))
+                        continue;
+
+                    String destination = it->name();
+
+                    if (it->isFile())
+                    {
+                        if ((*settings)[MergeTreeSetting::always_use_copy_instead_of_hardlinks])
+                        {
+                            mutation_context->new_part->getDataPartStorage().copyFileFrom(
+                                mutation_context->source_part->getDataPartStorage(), it->name(), destination);
+                        }
+                        else
+                        {
+                            mutation_context->new_part->getDataPartStorage().createHardLinkFrom(
+                                mutation_context->source_part->getDataPartStorage(), it->name(), destination);
+                        }
+                    }
+                }
+
+                mutation_context->new_part->checksums = mutation_context->source_part->checksums;
+
+                builder = std::make_unique<QueryPipelineBuilder>(std::move(mutation_context->mutating_pipeline_builder));
+
+                mutation_context->out = std::make_shared<MergedColumnOnlyOutputStream>(
+                    mutation_context->new_part,
+                    mutation_context->metadata_snapshot,
+                    mutation_context->updated_header.getNamesAndTypesList(),
+                    MergeTreeIndices{},
+                    ColumnsStatistics{},
+                    ctx->data->getContext()->chooseCompressionCodec(0, 0),
+                    mutation_context->source_part->index_granularity,
+                    mutation_context->source_part->getBytesUncompressedOnDisk());
+
+                mutation_context->mutating_pipeline = QueryPipelineBuilder::getPipeline(std::move(*builder));
+                mutation_context->mutating_pipeline.disableProfileEventUpdate();
+                mutation_context->mutating_executor = std::make_unique<PullingPipelineExecutor>(mutation_context->mutating_pipeline);
+            }
```

Comment:
This quite a huge part of the code looks like an exact copy paste from `MutateAllPartColumnsTask`. Could you move this to one place (maybe to `ProjectionMutationContext`).

Reply:
Good idea, I have moved all similar code into the struct. Much clean now.